x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/tool/BootstrapPasswordAndEnrollmentTokenForInitialNode.java

```diff
@@ -0,0 +1,187 @@
+    CommandLineHttpClient getClient(Environment env) {
+        return clientFunction.apply(env);
+    }
+    EnrollmentTokenGenerator getEnrollmentTokenGenerator(Environment env) throws Exception{
+        return createEnrollmentTokenFunction.apply(env);
+    }
+    KeyStoreWrapper getKeyStoreWrapper(Environment env) throws Exception { return keyStoreFunction.apply(env); }
+    OptionParser getParser() { return parser; }
+
+    BootstrapPasswordAndEnrollmentTokenForInitialNode() {
+        this(
+            environment -> new CommandLineHttpClient(environment),
+            environment -> KeyStoreWrapper.load(environment.configFile()),
+            environment -> new EnrollmentTokenGenerator(environment)
+        );
+    }
+
+    BootstrapPasswordAndEnrollmentTokenForInitialNode(Function<Environment, CommandLineHttpClient> clientFunction,
+                                                      CheckedFunction<Environment, KeyStoreWrapper, Exception> keyStoreFunction,
+                                                      CheckedFunction<Environment, EnrollmentTokenGenerator, Exception>
+                                                     createEnrollmentTokenFunction){
+        super("Set elastic password and generate enrollment token for initial node");
+        this.clientFunction = clientFunction;
+        this.keyStoreFunction = keyStoreFunction;
+        this.createEnrollmentTokenFunction = createEnrollmentTokenFunction;
+        parser.allowsUnrecognizedOptions();
+        keyStorePassword = parser.accepts("keystore-password", "keystore password").withRequiredArg();
+        docker = parser.accepts("docker", "determine that we are running in docker");
+    }
+
+    public static void main(String[] args) throws Exception {
+        exit(new BootstrapPasswordAndEnrollmentTokenForInitialNode().main(args, Terminal.DEFAULT));
+    }
+
+    @Override
+    protected void execute(Terminal terminal, OptionSet options, Environment env) throws Exception {
+        if (!options.has(keyStorePassword)) {
+            throw new UserException(ExitCodes.USAGE, null);
+        }
+
+        final SecureString keystorePassword = new SecureString(keyStorePassword.values(options).get(0));
+        SecureString password;
+        final CommandLineHttpClient client = getClient(env);
+        final EnrollmentTokenGenerator enrollmentTokenGenerator = getEnrollmentTokenGenerator(env);
+        keyStoreWrapper = getKeyStoreWrapper(env);
+        final SecureString bootstrapPassword = readBootstrapPassword(env, keystorePassword);
+        try {
+            client.checkClusterHealthWithRetriesWaitingForCluster(elasticUser, bootstrapPassword, 5);
+        } catch (Exception e) {
+            throw new UserException(ExitCodes.UNAVAILABLE, null);
+        }
+        if (ReservedRealm.BOOTSTRAP_ELASTIC_PASSWORD.exists(env.settings())) {
+            password = bootstrapPassword;
+        } else {
+            password = setElasticUserPassword(client, bootstrapPassword);
```

Comment:
what is the expected behavior if we can't set the elastic user password ? Should we fail and throw as we do now, or should we attempt to proceed with generating the tokens as we do when bootstrap.password is defined ?

Reply:
After moving the password setting at the end, we will always generate and print out the token

Reply:
I was mostly interested in if this has been already discussed and what we _think_ we should to be doing.

> After moving the password setting at the end, we will always generate and print out the token

This is not necessary. We can select to do otherwise, if we think it makes sense